### DISCUSSION TITLE
Ability to import source as raw/url

### DISCUSSION THREAD

**Discussion by maxsbelt** [web-infra-dev/rsbuild#2665] [Q&A] [answered]
I have a form where user should upload file in `.csv` format and there is an example of such file. So I need to:
- display the content of this file in the page (e.g. load it as a string with `/path/to/file/filename.csv?raw`)
- display the link to the file example for downloading (e.g. load it as a url with `/path/to/file/filename.csv?url`)

It works in `vite` and I saw something similar in the documentation:
https://www.rspack.dev/config/module#ruleresourcequery

But seems like it doesn't work:
|raw|url|
|-|-|
|<img width="659" alt="Screenshot 2024-06-21 at 18 24 14" src="https://github.com/web-infra-dev/rsbuild/assets/950216/da10a2bd-49c5-43e8-9b75-31930af12034">|<img width="667" alt="Screenshot 2024-06-21 at 18 25 09" src="https://github.com/web-infra-dev/rsbuild/assets/950216/2beb9f7a-6600-4554-8bc8-cbceaa1f9e0b">|

Is it supported?

**Comment by chenjiahan** [comment, marked as answer]
Try this:

```ts
// rsbuild.config.ts
export default defineConfig({
  tools: {
    rspack: {
      module: {
        rules: [
          {
            resourceQuery: /url$/,
            type: 'asset/resource',
          },
          {
            resourceQuery: /raw$/,
            type: 'asset/source',
          },
        ],
      },
    },
  },
});
```

**Reply by maxsbelt** [reply]
That works, thanks!

**Comment by ulivz** [comment]
@chenjiahan Thanks for this solution, when I configure it like this and try to import a `tsx`, I encounter `react/jsx-dev-runtime` injected into the bundled code. Is there a possible solution here?

Expected raw code:

```tsx
// src/hello-world.tsx
export const HelloWorld = () => {
  return <div>Hello world</div>
};
```

Usage:

```tsx
// src/index.tsx
import Source from './hello-world?raw';
```

value of `Source`:

```tsx
import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";
export const HelloWorld = ()=>{
    return /*#__PURE__*/ _jsxDEV("div", {
        children: "Hello world"
    }, void 0, false, {
        fileName: "...",
        lineNumber: 2,
        columnNumber: 10
    }, this);
};
_c = HelloWorld;
var _c;
$RefreshReg$(_c, "HelloWorld");

function $RefreshSig$() {
  return $ReactRefreshRuntime$.createSignatureFunctionForTransform();
}
function $RefreshReg$(type, id) {
  $ReactRefreshRuntime$.register(type, __webpack_module__.id + "_" + id);
}
Promise.resolve().then(function() {
  $ReactRefreshRuntime$.refresh(__webpack_module__.id, __webpack_module__.hot);
});
```

**Reply by chenjiahan** [reply]
This is indeed an issue that needs to be addressed. I will add an option in `@rspack/plugin-react-refresh` to allow excluding certain files based on resourceQuery.

Related code: https://github.com/rspack-contrib/rspack-plugin-react-refresh/blob/main/src/index.ts#L82-L91

**Reply by chenjiahan** [reply]
Try this:

1. Update `@rsbuild/plugin-react` to v1.3.1

2. Add custom config to `rsbuild.config.ts`:

```js
import { defineConfig } from '@rsbuild/core';
import { pluginReact } from '@rsbuild/plugin-react';

export default defineConfig({
  plugins: [
    pluginReact({
      reactRefreshOptions: {
        resourceQuery: { not: /raw/ },
      },
    }),
  ],
  tools: {
    bundlerChain(chain, { CHAIN_ID }) {
      chain.module.rule('source').resourceQuery(/raw$/).type('asset/source');
      chain.module.rule(CHAIN_ID.RULE.JS).resourceQuery({ not: /raw/ });
    },
  },
});
```

The above configuration is still a bit complicated and can be used as a workaround. I am planning to add built-in `?raw` query support for JS/TS modules in Rsbuild core.

**Comment by chenjiahan** [comment]
Rsbuild v1.4.0-beta.1 now provides built-in support for `.tsx?raw`, see:

- Release note: https://github.com/web-infra-dev/rsbuild/releases/tag/v1.4.0-beta.1
- Documentation: https://rsbuild.rs/guide/basic/static-assets#import-as-string
